### PR TITLE
Storage visibility v2

### PR DESCRIPTION
**Pull request by MCoulombe727**
I closed the other PR and opened one on the V2 branch as recommended. @Sytten how stable is the alpha branch? When do you think we could bump the version to an official 2.0.0?

### CODE REVIEW

File: src/main/java/com/coveo/spillway/storage/LimitUsageStorage.java [hunk 1]
```diff
@@ -120,6 +120,28 @@
   Map<LimitKey, Integer> debugCurrentLimitCounters();
 
   /**
+   * @param resource The resource for which you want to get the current limit counts
+   * @return A Map of all the enforced limits for a resource and their current count
+   *
+   */
+  Map<LimitKey, Integer> getCurrentLimitCounters(String resource);
```
Comment (malaporte) [inline src/main/java/com/coveo/spillway/storage/LimitUsageStorage.java]: Wouldn't it had been better to have only calls taking in all required stuff and returning an `Integer`? Returning the data as a Map looks a bit raw to me...

Comment (MCoulombe727) [inline src/main/java/com/coveo/spillway/storage/LimitUsageStorage.java]: You'd need to store all the keys or properties associated with a resource in a duplicated storage somewhere to be able to retrieve your limits for a resource. You'd also have to do a call for every individual limits which seems really inefficient.

Comment (malaporte) [inline src/main/java/com/coveo/spillway/storage/LimitUsageStorage.java]: But how do you suggest using the map that is returned? If you want to look up a specific limit, you'll need to build a `LimitKey` with a particular time bucket...

Comment (MCoulombe727) [inline src/main/java/com/coveo/spillway/storage/LimitUsageStorage.java]: You can easily search by the key's attributes. There's no need to recreate an entire LimitKey to consume what's returned. "getCurrentLimitCount(String resource, String limitName, String property)" could return an Integer as it'll realistically only ever have 1 key, but I preferred consistency.

**Comment by Sytten** [comment]
Nice! I will take a more thoughtful look into the changes made in this PR tomorrow after work, but starting with V2 is better.
We had this problem that we wanted to check if one limit was exceeded before increasing them (ex: if you a 10s limit and 5min limit you don`t want to trigger the 5min limit because you always exceed your 10s limit) and the way to do that was to increase by 0 and then increase by the number.
That could be improved before releasing V2 .

If you want to release a V2, we have to be sure of the interfaces since we won't be able to change them for some time (unless we bump to V3). I feel that its not stable enough (since we are still modifying them). It would be nice to make part of the technical document I wrote for the lib public so we could have a common roadmap for the project (only the part relevant to the lib and not the actual throttling plan of coveo).

File: src/main/java/com/coveo/spillway/storage/LimitUsageStorage.java [hunk 2]
```diff
@@ -120,6 +120,28 @@
   Map<LimitKey, Integer> debugCurrentLimitCounters();
 
   /**
+   * @param resource The resource for which you want to get the current limit counts
+   * @return A Map of all the enforced limits for a resource and their current count
```
Comment (jebeaudet) [inline src/main/java/com/coveo/spillway/storage/LimitUsageStorage.java]: I'd mention it's an unmodifiable instance on the three methods

File: src/main/java/com/coveo/spillway/storage/LimitUsageStorage.java [hunk 3]
```diff
@@ -120,6 +120,28 @@
...
+   */
+  Map<LimitKey, Integer> getCurrentLimitCounters(String resource, String limitName);
+
+  /**
+   * @param resource The resource for which you want to get the current limit counts
+   * @param limitName The limit name for which you want to get the current limit counts
+   * @param property The property for which you want to get the current limit counts
+   * @return A Map containing a specific limit for a resource and its current count
+   */
+  Map<LimitKey, Integer> getCurrentLimitCount(String resource, String limitName, String property);
```
Comment (jebeaudet) [inline src/main/java/com/coveo/spillway/storage/LimitUsageStorage.java]: `getCurrentLimitCounters`

File: src/main/java/com/coveo/spillway/storage/AsyncBatchLimitUsageStorage.java [hunk 4]
```diff
@@ -91,6 +91,22 @@ public AsyncBatchLimitUsageStorage(
   }
 
   @Override
+  public Map<LimitKey, Integer> getCurrentLimitCounters(String resource) {
+    return wrappedLimitUsageStorage.getCurrentLimitCounters(resource);
```
Comment (jebeaudet) [inline src/main/java/com/coveo/spillway/storage/AsyncBatchLimitUsageStorage.java]: unmodifiable? or the underlying one is already unmodifiable?

Comment (MCoulombe727) [inline src/main/java/com/coveo/spillway/storage/AsyncBatchLimitUsageStorage.java]: The wrapped storage already returns an instance of UnmodifiableMap

**Comment by pastjean** [comment]
GOOD job captain :D you can merge v2 into master after this. and delete all foreign branches ?

**Comment by Sytten** [comment]
Il faudrait faire une release sur maven central @pastjean

**Comment by pastjean** [comment]
Done ca prend du temps 2.0.0-alpha2 je pense

**Comment by pastjean** [comment]
http://search.maven.org/#search%7Cga%7C1%7Cg%3A%22com.coveo%22

**Comment by Sytten** [comment]
Ha ok cool, juste la badge qui est pas update à cause que c'est pas un stable

**Comment by Sytten** [comment]
@MCoulombe727 tu pourrais update la doc du gh-pages aussi ;)

**Pull request merged**